File: Continuous Subarray Sum.cpp
class Solution {
public:
    bool checkSubarraySum(vector<int>& nums, int k) {
        map <int, int> valueAt;
        int sum = nums[0] % k;
        valueAt[sum] = 0;
        for(int i = 1; i < (int) nums.size(); ++i){
            sum += nums[i];
            sum %= k;
            if(sum == 0 or valueAt.find(sum) != valueAt.end()){
                if(sum == 0 or valueAt[sum] < i - 1){
                    return true;
                }
            }
            if(valueAt.find(sum) == valueAt.end()){
                valueAt[sum] = i;
            }
        }
        return false;
    }
};

File: Continuous Subarray Sum.java
class Solution {
    public boolean checkSubarraySum(int[] nums, int k) {
        int currentSum = 0;
        HashMap <Integer, Integer> cumulativeSumAt = new HashMap <>();
        for(int index = 0; index < nums.length; ++index){
            currentSum += nums[index];
            currentSum %= k;
            if(cumulativeSumAt.containsKey(currentSum)){
                if(index - cumulativeSumAt.get(currentSum) > 1){
                    return true;
                }
            }
            else{
                cumulativeSumAt.put(currentSum, index);
            }
            if(currentSum == 0 && index > 0){
                return true;
            }
        }
        return false;
    }
}
